README.md
# Hierarchical Inference Testing

[![CRAN](http://www.r-pkg.org/badges/version/hit)](https://cran.r-project.org/package=hit)

The current built and test status for Linux (Mac)
[![Build Status](https://travis-ci.org/QTCAT/hit.svg)](https://travis-ci.org/QTCAT/hit)
and for Windows 
[![Build status](https://ci.appveyor.com/api/projects/status/kttq4x98q6hra6ct?svg=true)](https://ci.appveyor.com/project/jrklasen/hit)
.

## Description:
Hierarchical inference testing (HIT) for (generalized) linear models with 
correlated covariates. HIT is furthermore applicable to high-dimensional 
settings. For details see:

**Mandozzi, J. and Buehlmann, P. (2015)**. *Hierarchical testing in the 
high-dimensional setting with correlated variables*. Journal of the American 
Statistical Association. Preprint 
[arXiv:1312.5556](https://arxiv.org/abs/1312.5556).

## Installation:
The package can be installed from CRAN,

```R
install.packages("hit")

```

or via 
[`devtools`](https://github.com/hadley/devtools#updating-to-the-latest-version-of-devtools)
, if you haven't `devtools` installed yet you have to do so first.

```R
# install.packages("devtools")
devtools::install_github("QTCAT/hit")
```

## Example:
The `hit`-function example gives an overview of the functionality of the 
package and can be accessed once the package is loaded.

```R
library(hit)
example(hit)
```

--------------------------------------------------------------------------------
[![License](https://img.shields.io/badge/license-GPL%20%28%3E=%202%29-brightgreen.svg)](https://www.gnu.org/licenses/gpl-2.0.html)
&copy; 2016 JR Klasen


NEWS.md
## hit v0.4.0 (Release date: 2016-xx-xx)

Changes: 

* ``hit``: second path for lambda estimation, with a new lambda for each sample split.

## hit v0.3.1 (Release date: 2016-09-15)

Changes: 

* ``reorder``: a more general reorder function for hierarchy



## hit v0.3.0 (Release date: 2016-08-29)

Changes:

* ``hit``: remove the alpha optimization and allow only one alpha value

* ``hit``: allow the response variable (y) to be Poisson distributed

* ``summary.hit``: bug fix  



## hit v0.2-2 (Release date: 2016-04-26)

Changes:

* ``hit``: bug fixes in summary and some other small fixes



## hit v0.2-1 (Release date: 2016-01-11)

Changes:

* ``hit``: sticks to cross-validation as selection method

* ``hit``: arguments have changed



## hit v0.2-0 (Release date: 2015-12-04)

Changes:

* ``fast.anova``: a new ``fast.glmanova`` method for GLM's

* ``hit``: now is able to deal with binomial responses

* ``hit``: a new selection method for the active set


tests/testthat/test_hit.R
context("HIT testing")

test_that("hit testing", {
  set.seed(123)
  n <- 100
  p <- 150
  # x with correlated columns
  corMat <- toeplitz((p:1 / p)^3)
  corMatQ <- chol(corMat)
  x <- matrix(rnorm(n * p), nrow = n) %*% corMatQ
  colnames(x) <- paste0("x", 1:p)
  # y
  y <- x[, c(3, 5, 73)] %*% c(2, 5, 3) + rnorm(n)
  # hierarchy
  dend <- as.dendrogram(hclust(dist(t(x))))
  hier <- as.hierarchy(dend)
  # HIT
  fit <- hit(x, y, hier)
  # checks
  expect_equal(class(fit), "hit")
  expect_equal(unname(unlist(lapply(fit, class))), 
               c("numeric", "numeric","hierarchy", 
                 "logical", "numeric", "numeric"))
  expect_equal(names(fit), 
               c("pValues", "selectFreq", "hierarchy", 
                 "tested", "alpha", "lambda"))
})


tests/testthat/test_hierarchy.R
context("hirarchy testing")

test_that("hirarchy testing", {
  set.seed(123)
  n <- 100
  p <- 150
  # x with correlated columns
  corMat <- toeplitz((p:1/p)^3)
  corMatQ <- chol(corMat)
  x <- matrix(rnorm(n * p), nrow = n) %*% corMatQ
  colnames(x) <- paste0("x", 1:p)
  # y
  y <- x[, c(3, 5, 73)] %*% c(2, 5, 3) + rnorm(n)
  # hierarchy
  hc <- hclust(dist(t(x)))
  hier <- as.hierarchy(hc, max.height = 20)
  # check:
  expect_equal(class(hier), "hierarchy")
#   expect_equal(unname(unlist(lapply(hier, class))), 
#                c("list", "list", "character", "numeric"))
#   expect_equal(names(hier), 
#                c("hierarchyCluster", "clusterMembers", 
#                  "labels", "hierarchyLevel"))
})


R/anova.R
#' @title Fast ANOVA
#' 
#' @description A fast sequential analysis of variance (ANOVA). Mainly 
#' developed for internal use.
#' 
#' @param x Design matrix of dimension \code{n * p}.
#' @param y Response vector of observations of length \code{n}.
#' @param assign Integer vector assigning columns to terms can be also given as 
#' \code{x} attribute in which case the argument is ignored. If an intercept exist it is 
#' expected to be the first column in \code{x} and it has to be specified by a '0' in tis 
#' vector. For details about assign see 
#' \code{\link[stats]{model.matrix}}.
#' @param family A description of the error distribution and link function to be used in 
#' the model. For glm this can be a character string naming a family function or the result 
#' of a call to a family function. (See \code{\link[stats]{family}} for details of family 
#' functions.)
#' @param test The name of the test either "LRT" (default) for likelihood ratio test or "F" 
#' for F test.
#' 
#' @seealso \code{\link[stats]{lm}}, \code{\link[stats]{anova}}, and 
#' \code{\link[stats]{aov}}.
#' 
#' @examples
#' y <- rnorm(n = 100)
#' x <- matrix(data = rnorm(1000), nrow = 100)
#' a <- 1:10
#' fast.anova(x = x, y = y, assign = a)
#' 
#' @importFrom stats binomial gaussian Gamma inverse.gaussian poisson quasi quasibinomial 
#' quasipoisson
#' @export 
fast.anova <- function(x, y, assign = NULL, family = gaussian(), 
                       test = c("LRT", "F")) {
  ##### check validity of args
  if (is.null(assign) && !is.null(attr(x, "assign"))) {
    assign <- attr(x, "assign")
  }
  if (is.null(assign)) {
    stop("either a 'x' attribute 'assign' or the 'assign' argument must be specified")
  }
  stopifnot(ncol(x) == length(assign))
  stopifnot(nrow(x) == length(y))
  if (is.character(family)) {
    family <- eval(call(family))
  } else if (is.function(family)) {
    family <- family()
  }
  test <- match.arg(test, c("LRT", "F"), several.ok = TRUE)
  # fit anova model
  if (family$family == "gaussian" && any(test == "F")) 
    p <- fast.lmanova(x, y, assign)
  else
    p <- fast.glmanova(x, y, assign, family, test[1])
  p
}


#' @title Fast LM F test ANOVA
#' 
#' @description A fast sequential analysis of variance (ANOVA). Mainly developed for 
#' internal use.
#' 
#' @param x Design matrix of dimension \code{n * p}.
#' @param y Response vector of observations of length \code{n}.
#' @param assign Integer vector assigning columns to terms.
#' 
#' @importFrom stats lm.fit pf 
#' @keywords internal
fast.lmanova <- function(x, y, assign) {
  # LM fit by pivoted QR decomposition
  fit <- lm.fit(x, y)
  if (assign[1L] == 0L) { 
    ##### with intercept
    full.rank <- 1L:(fit$rank - 1L)
    assign.pivot <- assign[fit$qr$pivot[full.rank + 1L]]
    var <- fit$effects[-1L]^2
  } else { 
    ##### without intercept
    full.rank <- 1L:fit$rank
    assign.pivot <- assign[fit$qr$pivot[full.rank]]
    var <- fit$effects^2
  }
  # Treatment: Sum Sq | DF | Mean Sq
  ss.treat <- tapply(var[full.rank], assign.pivot, "sum")
  df.treat <- table(assign.pivot)
  ms.treat <- ss.treat / df.treat
  # Residuals: Sum Sq | DF | Mean Sq
  ss.res <- sum(var[-full.rank])
  df.res <- fit$df.residual
  ms.res <- ss.res / df.res
  # F value
  f <- ms.treat / ms.res 
  # p value
  p <- rep(1, max(assign))
  p[unique(assign.pivot)] <- pf(f, df.treat, df.res, lower.tail = FALSE)
  p
}


#' @title Fast GLM F test of LR test ANOVA
#' 
#' @description A fast sequential analysis of variance (ANOVA). Mainly developed for 
#' internal use.
#' 
#' @param x Design matrix of dimension \code{n * p}.
#' @param y Response vector of observations of length \code{n}.
#' @param assign Integer vector assigning columns to terms.
#' @param family A description of the error distribution and link function to be used in 
#' the model. For glm this must be the result of a call to a family function.
#' @param test The name of the test either "LRT" (default) for likelihood ratio  test or 
#' "F" for F test.
#' 
#' @importFrom speedglm speedglm.wfit
#' @importFrom stats pf pchisq
#' @keywords internal
fast.glmanova <- function(x, y, assign, family, test) {
  inter <- ifelse(assign[1] == 0L, TRUE, FALSE)
  # Full model fit by pivoted Colesky decomposition
  suppressWarnings(
    full.fit <- speedglm.wfit(X = x, y = y, 
                              intercept = inter, family = family, 
                              method = "Cholesky", tol.solve = 1e-30)
  )
  # Reduced model fits by pivoted Colesky decomposition
  red.dev <- red.df <- NULL 
  if (max(assign) > 1L) {
    for (i in seq_len( max(assign) - 1L)) {
      suppressWarnings(
        red.fit <- speedglm.wfit(X = x[, assign <= i, drop = FALSE], y = y, 
                                 intercept = inter, family = family, 
                                 method = "Cholesky", tol.solve = 1e-30)
      )
      red.dev <- c(red.dev, red.fit$deviance)
      red.df <- c(red.df, red.fit$df)
    }
  }
  # Treatment: Variance | DF
  df.treat <- -diff(c(full.fit$nulldf, red.df, full.fit$df))
  var.treat <- pmax(-diff(c(full.fit$nulldev, red.dev, full.fit$deviance)))
  ##### check for singularity
  nonsing.covar <- which(df.treat != 0)
  df.treat <- df.treat[nonsing.covar]
  var.treat <- var.treat[nonsing.covar]
  ##### Dispersion factor
  disp <- full.fit$dispersion
  # estimated  test statistic
  p <- rep(1, max(assign))
  if (test == "F") {
    #  residuals: DF
    df.res <- ifelse(disp == 1, Inf, full.fit$df)
    # F value
    f <- var.treat / (disp * df.treat)
    # p value
    p[nonsing.covar] <- pf(f, df.treat, df.res, lower.tail = FALSE)
  } else {
    # Chi value
    chi <- var.treat / disp
    # p value
    p[nonsing.covar] <- pchisq(chi, df.treat, lower.tail = FALSE)
  }
  p
}


src/hit.cpp
#include <Rcpp.h>

using namespace Rcpp;



// find significant cluster from p-value matrix
// [[Rcpp::export]]
NumericMatrix sigCluster(NumericVector pVal, List clustLevel, List clustMemb,
                         const int p, const double alpha, 
                         int minDistInx, int maxDistInx) 
{
    // R Indices to C++
    minDistInx --;
    maxDistInx --;
    // Initialisation
    NumericMatrix sigClust(2, p);
    for (int l = 0; l < p; ++ l) {
        sigClust(1, l) = 1;
    }
    IntegerVector gIDs;
    std::vector<int> gInx;
    int k = 1;
    int m, h;
    bool newClust = true;
    bool isZero = false;
    bool isNotZero = false;
    // start searching between min and max distance
    for (int i = minDistInx; i > maxDistInx; -- i) {
        gIDs = clustLevel[i];
        // test each cluster at dist i
        for (int j = 0; j < gIDs.size(); ++ j) {
            m = gIDs[j] -1; // R Indix to C++
            gInx = clustMemb[m]; 
            // test if all variables of cluster have already been significant 
            // before at smaller distances
            for (int l = 0; l < gInx.size(); ++ l) {
                h = gInx[l] -1; // R Indix to C++
                if (sigClust(0, h) == 0) {
                    isZero = true;
                }
                if (sigClust(0, h) != 0) {
                    isNotZero = true;
                }
                // is variable already part of sig cluster?
                // or p-value is too large jump to next cluster
                if ((isZero && isNotZero) || (pVal[m] > alpha)) {
                    newClust = false;
                    break;
                }
            }
            // if newClust == true make sig cluster
            if (newClust) {
                for(int l = 0; l < gInx.size(); ++ l) {
                    h = gInx[l] -1; // R Indix to C++
                    sigClust(0, h) = k;
                    sigClust(1, h) = pVal[m];
                }
                ++ k;
            }
            newClust = true;
            isZero = false;
            isNotZero = false;
            gInx.clear();  
        }
    }
    // ordert cluster index 
    NumericVector cInx = sort_unique(sigClust(0, _));
    for (int i = 0; i < cInx.size(); ++ i) {
        for (int l = 0; l < p; ++ l) {
            if (sigClust(0, l) == cInx[i]) {
                sigClust(0, l) = i;
            }
        }
    }
    return sigClust;
} // end sigCluster()


R/hit.R
#' @title Hierarchical Inference Testing
#'
#' @description Hierarchical inference testing for linear models with high-dimensional 
#' and/or correlated covariates by repeated sample splitting.
#'
#' @param x Design matrix of dimension \code{n * p}, without intercept. Variables not part 
#' of the dendrogram are added to the HO-model, see Details below.
#' @param y Quantitative response variable dimension \code{n}.
#' @param hierarchy Object of class \code{\link{as.hierarchy}}. Must include all variables 
#' of \code{x} which should be tested.
#' @param family Family of response variable distribution. Ether \code{y} is "gaussian" or 
#' "poisson" in which case \code{y} must be a vector or it is "binomial" distibuded and is 
#' either a vecror of zeros and ones, factor with two levels,  or a two-column matrix of 
#' counts or proportions. The second column is treated as the target class. For a factor, 
#' the last level in alphabetical order is the target class. For "binomial" if \code{y} is 
#' presented as a vector, it will be coerced into a factor.
#' @param B Number of sample-splits.
#' @param p.samp1 Fraction of data used for the LASSO. The hierachical ANOVA testing uses 
#' \code{1 - p.samp1}.
#' @param nfolds Number of folds (default is 10). See \code{\link[glmnet]{cv.glmnet}} for 
#' more details.
#' @param overall.lambda Logical, if true, lambda is estimated once, if false, lambda is 
#' estimated for each sample split.
#' @param lambda.opt Criterion for optimum selection of cross-validated lasso. Either 
#' "lambda.1se" (default) or "lambda.min". See \code{\link[glmnet]{cv.glmnet}} for more 
#' details.
#' @param alpha A single value in the range of 0 to 1 for the elastic net mixing parameter.
#' @param gamma Vector of gamma-values.
#' @param max.p.esti Maximum alpha level. All p-values above this value are set to one. 
#' Small \code{max.p.esti} values reduce computing time.
#' @param mc.cores Number of cores for parallelising. Theoretical maximum is 'B'. For 
#' details see \code{\link[parallel]{mclapply}}.
#' @param trace If TRUE it prints current status of the program.
#' @param ... Additional arguments for \code{\link[glmnet]{cv.glmnet}}.
#'
#' @details The H0-model contains variables, with are not tested, like experimental-design 
#' variables. These variables are not penalised in the LASSO model selection and are always 
#' include in the reduced ANOVA model.
#'
#' @references Mandozzi, J. and Buehlmann, P. (2013). \emph{Hierarchical testing in the 
#' high-dimensional setting with correlated variables}. To appear in the Journal of the 
#' American Statistical Association. Preprint arXiv:1312.5556
#'
#' @examples
#'
#' # Simulation:
#' set.seed(123)
#' n <- 80
#' p <- 82
#' ## x with correlated columns
#' corMat <- toeplitz((p:1/p)^2)
#' corMatQ <- chol(corMat)
#' x <- matrix(rnorm(n * p), nrow = n) %*% corMatQ
#' colnames(x) <- paste0("x", 1:p)
#' ## y
#' mu <- x[, c(5, 10, 72)] %*% c(2, -2, 2)
#' y <-  rnorm(n, mu)
#' ## clustering of the clumns of x
#' hc <- hclust(dist(t(x)))
#'
#' # HIT with AF
#' out <- hit(x, y, hc)
#' summary(out)
#'
#' @importFrom parallel mclapply
#' @importFrom glmnet cv.glmnet
#' @importFrom stats reorder
#' @export
hit <- function(x, y, hierarchy, family = "gaussian", B = 50, p.samp1 = 0.5, 
                nfolds = 10, overall.lambda = FALSE, lambda.opt = "lambda.1se", alpha = 1,
                gamma = seq(0.05, 0.99, length.out = 100), max.p.esti = 1,
                mc.cores = 1L, trace = FALSE, ...) {
  #   Mandozzi and Buehlmann (2015), 2 Description of method
  ## Checks args
  if (is.null(colnames(x)))
    stop("column names of 'x' are missing")
  n <- nrow(x)
  p <- ncol(x)
  if (class(hierarchy) == "dendrogram" || class(hierarchy) == "hclust")
    hierarchy <- as.hierarchy(hierarchy)
  stopifnot(class(hierarchy) == "hierarchy")
  lambda.opt <- match.arg(lambda.opt, c("lambda.1se", "lambda.min"))
  ##### Check family and assign test
  family <- match.arg(family, c("gaussian", "binomial", "poisson"))
  test <- "LRT"
  if (family == "gaussian")
    test <- "F"
  family2 <- eval(call(family))
  ##### Checks order of variats
  x.names <- colnames(x)
  hier.names <- names(hierarchy)
  if (!all(hier.names %in% x.names))
    stop("'hierarchy' includs variabels not in 'x'")
  if (identical(hier.names, x.names)) {
    x.notest <- integer(0L)
  } else if (setequal(hier.names, x.names)) {
    hierarchy <- reorder(hierarchy, x.names)
    x.notest <- integer(0L)
  } else {
    hierarchy <- reorder(hierarchy, x.names)
    x.notest <- match(setdiff(x.names, names(hierarchy)), x.names)
  }
  penalty.factor <- rep(1L, p)
  penalty.factor[x.notest] <- 0L
  ##### Sub-sample splits
  if (p.samp1 < .1 | p.samp1 > .9)
    stop("'p.samp1' must be between .1 and 0.9")
  n.samp1 <- as.integer(n * p.samp1)
  n.samp2 <- n - n.samp1
  allSamp1.ids <- replicate(B, sample.int(n, n.samp1), simplify = FALSE)
  ##  2.2 Screening
  if (trace)
    cat("LASSO has started at:\n\t", as.character(Sys.time()), "\n")
  if (overall.lambda) {
    lambdas <- opt.penalty(x, y, family, alpha, lambda.opt, nfolds,
                             penalty.factor, n.samp2, ...)
    allActSet.ids <- mclapply(allSamp1.ids, samp1.lasso.overall,
                              x, y, family, alpha, lambdas,
                              penalty.factor, n.samp2, ...,
                              mc.cores = mc.cores, mc.cleanup = TRUE)
  } else {
    allActSet.ids <- mclapply(allSamp1.ids, samp1.lasso.split,
                              x, y, family, alpha, lambda.opt, nfolds,
                              penalty.factor, n.samp2, ...,
                              mc.cores = mc.cores, mc.cleanup = TRUE)
  }

  ##  2.3 Testing and multiplicity adjustmen; and
  ##  2.4 Aggregating and Hierarchical adjustment
  if (trace)
    cat("Significance testing has started at:\n\t",
        as.character(Sys.time()), "\n")
  pValues <- samp2.sigHierarchy(1L, x, y, hierarchy, family2, test, x.notest,
                                B, allSamp1.ids, allActSet.ids, 0,
                                max.p.esti, sort(gamma), 0L, mc.cores)
  ##### Results
  if (trace)
    cat("HIT has finished at:\n\t", as.character(Sys.time()), "\n")
  asi <- sort(unlist(allActSet.ids))
  sel.tab <- rep(0, p)
  sel.tab[unique(asi)] <- table(asi) / B
  tested <- rep(TRUE, p)
  tested[x.notest] <- FALSE
  names(tested) <- x.names
  out <- list(pValues = pValues,
              selectFreq = sel.tab,
              hierarchy = hierarchy,
              tested = tested,
              max.p.esti = max.p.esti)
  class(out) <- "hit"
  out
}


#' @title Cross-validation of LASSO penenlty lambda
#'
#' @description Cross-validation of LASSO penenlty lambda.
#'
#' @param x Design matrix, of dimension n x p.
#' @param y Vector of quantitative response variable.
#' @param family Distribution family of \code{y}.
#' @param alpha A single value n the range of 0 to 1 for the elastic net mixing parameter. 
#' @param lambda.opt Criterion for optimum selection of cross-validated lasso. Either 
#' "lambda.1se" (default) or "lambda.min". See \code{\link[glmnet]{cv.glmnet}} for more 
#' details.
#' @param nfolds Number of folds (default is 10). See
#' @param penalty.factor See glmnet.
#' @param n.samp2 Number of individuals in samp2 which is the max. for non zero 
#' coefficients.
#' @param ... Additional agruments.
#'
#' @importFrom parallel mclapply
#' @importFrom glmnet cv.glmnet
#' @keywords internal
opt.penalty <- function(x, y, family, alpha, lambda.opt, nfolds, 
                        penalty.factor, n.samp2, ...) {
  foldid <- sample(rep(1L:nfolds, length.out = nrow(x)))
  suppressWarnings(
    cvfit <- cv.glmnet(x, y, family = family, foldid = foldid, alpha = alpha,
                       penalty.factor = penalty.factor, pmax = n.samp2 - 2L, ...)
  )
  if (lambda.opt == "lambda.min")
    optlambda <- cvfit$lambda[cvfit$lambda >= cvfit$lambda.min]
  else
    optlambda <- cvfit$lambda[cvfit$lambda >= cvfit$lambda.1se]
  optlambda
}



#' @title Variabel Screening
#'
#' @description LASSO function of the HIT algorithem.
#'
#' @param samp1 List of index for subsample (mclapply index).
#' @param x Design matrix, of dimension n x p.
#' @param y Vector of quantitative response variable.
#' @param family Distribution family of \code{y}.
#' @param alpha Mixing value for elnet.
#' @param lambda A vector of lambda values sorted from large to small where the smallest is 
#' the optimal value or NULL.
#' @param penalty.factor See glmnet.
#' @param n.samp2 Number of individuals in samp2 which is the max. for non zero 
#' coefficients.
#' @param ... Additional agruments.
#'
#' @importFrom glmnet glmnet cv.glmnet
#' @importFrom stats coef
#' @keywords internal
samp1.lasso.overall <- function(samp1, x, y, family, alpha, lambda,
                                penalty.factor, n.samp2, ...) {
  suppressWarnings(
    fit <- glmnet(x[samp1, ], y[samp1], family = family,
                  alpha = alpha, lambda = lambda,
                  penalty.factor = penalty.factor, pmax = n.samp2 - 2L, ...)
  )
  beta <- coef(fit, s = lambda[length(lambda)])[-1L]
  actSet <- which(beta != 0 & penalty.factor == 1L)
  actSet
}

#' @title Variabel Screening
#'
#' @description LASSO function of the HIT algorithem.
#'
#' @param samp1 List of index for subsample (mclapply index).
#' @param x Design matrix, of dimension n x p.
#' @param y Vector of quantitative response variable.
#' @param family Distribution family of \code{y}.
#' @param alpha Mixing value for elnet.
#' @param lambda.opt Criterion for optimum selection of cross-validated lasso. Either 
#' "lambda.1se" (default) or "lambda.min". See \code{\link[glmnet]{cv.glmnet}} for more 
#' details.
#' @param nfolds Number of folds (default is 10). See
#' @param penalty.factor See glmnet.
#' @param n.samp2 Number of individuals in samp2 which is the max. for non zero 
#' coefficients.
#' @param ... Additional agruments.
#'
#' @importFrom glmnet cv.glmnet
#' @importFrom stats coef
#' @keywords internal
samp1.lasso.split <- function(samp1, x, y, family, alpha, lambda.opt, nfolds, 
                              penalty.factor, n.samp2, ...) {
  suppressWarnings(
    fit <- cv.glmnet(x[samp1, ], y[samp1], family = family,  nfolds = nfolds, 
                     alpha = alpha, penalty.factor = penalty.factor, 
                     pmax = n.samp2 - 2L, ...)
  )
  if (lambda.opt == "lambda.min")
    beta <- coef(fit, s = fit$lambda.min)[-1L]
  else
    beta <- coef(fit, s = fit$lambda.1se)[-1L]
  actSet <- which(beta != 0 & penalty.factor == 1L)
  actSet
}


#' @title Variabel Testing along the hierarchy
#'
#' @description ANOVA Testing, Multiplicity Adjustment, Aggregating and Hierarchical 
#' Adjustment
#'
#' @param j Index for cluster (mclapply index).
#' @param x Design matrix, of dimension n x p.
#' @param y Vector of quantitative response variable.
#' @param hierarchy A hierarchy object.
#' @param family Distribution family of \code{y}.
#' @param test name of test.
#' @param x.notest  Vector of indeces of non tested variabels.
#' @param B Number of sample-splits.
#' @param allSamp1.ids  List of subsampels.
#' @param allActSet.ids List of active sets.
#' @param upper.p P value upper huerarchy level of the clusters variables.
#' @param max.p.esti Maximum alpha level. All p-values above this value are set to one. 
#' Small max.p.esti values reduce computing time.
#' @param gamma Vector of gamma-values.
#' @param hl.count Hierarchy level counter for parallelism.
#' @param cores Number of cores for parallelising.
#'
#' @importFrom stats quantile
#' @keywords internal
samp2.sigHierarchy <- function(j, x, y, hierarchy, family, test, x.notest,
                               B, allSamp1.ids, allActSet.ids, upper.p,
                               max.p.esti, gamma, hl.count, cores) {
  ## 2.3 Testing and multiplicity adjustment
  cluster <- hierarchy[[j]]
  p.cluster <- sapply(1L:B, samp2.sigNode,
                      x, y, cluster, family, test, x.notest,
                      allSamp1.ids, allActSet.ids)
  ##  2.4 Aggregating and Hierarchical adjustment
  ### 2.4-1 Aggregating
  q.aggre <- sapply(gamma,
                    function(i, x) { min(1, quantile(x / i, i)) },
                    x = p.cluster)
  p.aggre <- min(1, (1 - log(min(gamma))) * min(q.aggre))
  ### 2.4-2 Hierarchical adjustment
  p.value <- max(p.aggre, upper.p)
  ##### Estimation at next lower level and find a way to parallelize
  if (!is.null(js <- attr(cluster, "subset"))) {
    if (p.value < max.p.esti) {
      if (hl.count == 0L && length(js) >= cores) {
        hl.count <- as.integer(sqrt(cores))
        pValues <- mclapply(js, samp2.sigHierarchy,
                            x, y, hierarchy, family, test, x.notest,
                            B, allSamp1.ids, allActSet.ids, p.value,
                            max.p.esti, gamma, hl.count + 1L, cores,
                            mc.cores = cores)
      } else if (hl.count <= as.integer(sqrt(cores))) {
        mc.cores <- ifelse(as.integer(sqrt(cores)) == 1L, 1L, 2L)
        pValues <- mclapply(js, samp2.sigHierarchy,
                            x, y, hierarchy, family, test, x.notest,
                            B, allSamp1.ids, allActSet.ids, p.value,
                            max.p.esti, gamma, hl.count + 1L, cores,
                            mc.cores = mc.cores, mc.allow.recursive = TRUE)
      } else {
        pValues <- lapply(js, samp2.sigHierarchy,
                          x, y, hierarchy, family, test, x.notest,
                          B, allSamp1.ids, allActSet.ids, p.value,
                          max.p.esti, gamma, hl.count + 1L, cores)
      }
    } else {
      pOne <- function(j) {
        if (!is.null(js <- attr(hierarchy[[j]], "subset")))
          return(c(1, sapply(js, pOne)))
        return(1)
      }
      pValues <-  sapply(js, pOne)
    }
  } else {
    pValues <- c()
  }
  out <- c(p.value, unlist(pValues))
  out
}


#' @title ANOVA testing and multiplicity adjustment
#'
#' @description ANOVA test (at single node) of the HIT algorithem.
#'
#' @param k Index for subsample (mclapply index).
#' @param x Design matrix, of dimension n x p.
#' @param y Vector of quantitative response variable.
#' @param cluster Clusters to be tested.
#' @param family Distribution family of \code{y}.
#' @param test name of test.
#' @param x.notest Vector of indeces of non tested variabels.
#' @param allSamp1.ids  List of subsampels.
#' @param allActSet.ids List of active sets.
#'
#' @keywords internal
samp2.sigNode <- function(k, x, y, cluster, family, test, x.notest,
                          allSamp1.ids, allActSet.ids) {
  ## 2.3 Testing and multiplicity adjustment
  n <- nrow(x)
  actClust <- intersect(allActSet.ids[[k]], cluster)
  nonActClust <- setdiff(allActSet.ids[[k]], cluster)
  ### 2.3-1 Testing
  p.cluster <- 1
  if (l.actClust <- length(actClust)) {
    ##### ANOVA between active set and active set minus cluster
    samp2 <- (1L:n)[-allSamp1.ids[[k]]]
    if (is.matrix(y))
      y <- y[samp2, ]
    else
      y <- y[samp2]
    if (l.nonActClust <- length(nonActClust)) {
      if (l.nonTested <- length(x.notest)) {
        x  <- cbind(1L, x[samp2, c(x.notest, nonActClust, actClust)])
        assign <- rep(0L:3L, c(1L, l.nonTested, l.nonActClust, l.actClust))
        get.p <- 3L
      } else {
        x  <- cbind(1L, x[samp2, c(nonActClust, actClust)])
        assign <- rep(0L:2L, c(1L, l.nonActClust, l.actClust))
        get.p <- 2L
      }
    } else {
      if (l.nonTested <- length(x.notest)) {
        x  <- cbind(1L, x[samp2, c(x.notest, actClust)])
        assign <- rep(0L:2L, c(1L, l.nonTested, l.actClust))
        get.p <- 2L
      } else {
        x  <- cbind(1L, x[samp2, actClust])
        assign <- rep(0L:1L, c(1L, l.actClust))
        get.p <- 1L
      }
    }
    p.test <- fast.anova(x, y, assign, family, test)[get.p]
    ### 2.3-2 Multiplicity adjustment
    p.cluster <- min(1, p.test * (l.actClust + l.nonActClust) / l.actClust)
  }
  p.cluster
}


#' @title Summary of HIT
#'
#' @description Significant clusters at alpha threshold.
#'
#' @param object A \code{\link{hit}} object.
#' @param alpha A alpha significance threshold.
#' @param max.height max. Height to consider.
#' @param ... Further arguments passed to or from other methods (not used).
#'
#' @method summary hit
#' @export
summary.hit <- function(object, alpha = 0.05, max.height, ...) {
  make.pVal <- function(i) {
    p.value <- object$pValues[i]
    inx <- object$hierarchy[[i]]
    height <- attr(object$hierarchy[[i]], "height")
    if (p.value <= alpha && 
        (all(is.na(P.CLUSTER[inx])) || 
         (all(!is.na(P.CLUSTER[inx])) && 
          all(P.CLUSTER[inx] <= alpha)))) {
      P.CLUSTER[inx] <<-  p.value
      ID.CLUSTER[inx] <<- COUNTER
      H.CLUSTER[inx] <<- height
      COUNTER <<- COUNTER + 1L
    }
    return(NULL)
  }
  stopifnot(inherits(object, "hit"))
  stopifnot(alpha <= 0.999)
  if (missing(max.height))
    max.height <- attr(object$hierarchy[[1L]], "height")
  P.CLUSTER <- rep(NA_real_, length(object$hierarchy[[1L]]))
  ID.CLUSTER <- rep(NA_integer_, length(object$hierarchy[[1L]]))
  H.CLUSTER <- rep(NA_real_, length(object$hierarchy[[1L]]))
  COUNTER <- 1L
  lapply(length(object$hierarchy):1L, make.pVal)
  non.na <- which(!is.na(ID.CLUSTER))
  out <- data.frame(clusters = ID.CLUSTER[non.na],
                    heights = H.CLUSTER[non.na],
                    pValues = P.CLUSTER[non.na])
  rownames(out) <- names(object$hierarchy[[1L]])[non.na]
  out <- out[out$heights <= max.height, ]
  if (ll <- length(unique(out[, 1L])))
    out[, 1L] <- as.integer(factor(out[, 1L], labels = 1L:ll))
  out
}
